crates/holochain/src/conductor/handle/graft_records_onto_source_chain.rs

```diff
@@ -0,0 +1,394 @@
+use holochain_state::source_chain::SourceChain;
+use holochain_types::prelude::ChainItem;
+
+use super::*;
+
+pub(crate) async fn graft_records_onto_source_chain(
+    handle: Arc<ConductorHandleImpl>,
+    cell_id: CellId,
+    validate: bool,
+    records: Vec<Record>,
+) -> ConductorApiResult<()> {
+    // Get or create the space for this cell.
+    // Note: This doesn't require the cell be installed.
+    let space = handle.conductor.get_or_create_space(cell_id.dna_hash())?;
+
+    let network = handle
+        .conductor
+        .holochain_p2p()
+        .to_dna(cell_id.dna_hash().clone());
+
+    let source_chain: SourceChain = space
+        .source_chain(handle.keystore().clone(), cell_id.agent_pubkey().clone())
+        .await?;
+
+    let existing = source_chain
+        .query(ChainQueryFilter::new().descending())
+        .await?
+        .into_iter()
+        .map(|r| r.signed_action)
+        .collect::<Vec<SignedActionHashed>>();
+
+    let graft = ChainGraft::new(existing, records).rebalance();
+    let chain_top = graft.existing_chain_top();
+
+    if validate {
+        validate_records(handle.clone(), &cell_id, &chain_top, graft.incoming()).await?;
+    }
+
+    // Produce the op lights for each record.
+    let data = graft
+        .incoming
+        .into_iter()
+        .map(|el| {
+            let ops = produce_op_lights_from_records(vec![&el])?;
+            // Check have the same author as cell.
+            let (sah, entry) = el.into_inner();
+            if sah.action().author() != cell_id.agent_pubkey() {
+                return Err(StateMutationError::AuthorsMustMatch);
+            }
+            Ok((sah, ops, entry.into_option()))
+        })
+        .collect::<StateMutationResult<Vec<_>>>()?;
+
+    // Commit the records to the source chain.
+    let ops_to_integrate = space
+        .authored_db
+        .async_commit({
+            let cell_id = cell_id.clone();
+            move |txn| {
+                if let Some((_, seq)) = chain_top {
+                    // Remove records above the grafting position.
+                    //
+                    // NOTES:
+                    // - the chain top may have moved since the grafting call began,
+                    //   but it doesn't really matter, since we explicitly want to
+                    //   clobber anything beyond the grafting point anyway.
+                    // - if there is an existing fork, there may still be a fork after the
+                    //   grafting. A more rigorous approach would thin out the existing
+                    //   actions until a single fork is obtained.
+                    txn.execute(
+                        "DELETE FROM Action WHERE author = ? AND seq > ?",
```

Comment:
are we still trying to migrate sql into separate `.sql` files that can be audited by the linter?

Reply:
Trying -- this was essentially just something moved from one file to another. I can move this into a SQL file to start building the habit of refactoring these as we see them.